pombola/south_africa/views/download.py

```diff
@@ -0,0 +1,86 @@
+import os
+
+from django.db.models import Prefetch, Q
+from django.http import Http404, StreamingHttpResponse
+from django.shortcuts import get_object_or_404
+from django.views.generic import TemplateView
+
+import xlsx_streaming
+from pombola.core.models import Contact, Organisation, Person, Position
+
+MP_DOWNLOAD_TEMPLATE_SHEET = os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), "mp-download-template.xlsx"
+)
+
+
+def get_email_address_for_person(person):
+    if person.email:
+        return person.email
+    if len(person.email_addresses) > 0:
+        return person.email_addresses[0].value
+    return ""
+
+
+def download_members_xlsx(request, slug):
+    organisation = get_object_or_404(Organisation, slug=slug)
+
+    # Get all of the currently_active positions at the organisation
+    organisation_positions = (
+        Position.objects.currently_active()
+        .filter(organisation=organisation)
+        .values("id")
+    )
+
+    party_positions = (
+        Position.objects.currently_active()
+        .filter(title__slug="member")
+        .filter(organisation__kind__slug="party")
+        .select_related("organisation")
+    )
+
+    cell_phone_contacts = Contact.contact_number_contacts()
+    email_contacts = Contact.email_contacts()
+
+    # Get the persons from the positions
+    persons = (
+        Person.objects.filter(hidden=False)
+        .filter(position__id__in=organisation_positions)
+        .distinct()
+        .prefetch_related(
+            "alternative_names",
+            Prefetch(
+                "position_set",
+                queryset=party_positions,
+                to_attr="active_party_positions",
+            ),
+            Prefetch("contacts", queryset=cell_phone_contacts, to_attr="cell_numbers"),
+            Prefetch("contacts", queryset=email_contacts, to_attr="email_addresses"),
```

Comment:
very cool - so does it end up just doing one request for the persons, with a join for positions, cell numbers and email addresses?

Reply:
It does six large queries instead of many smaller queries, i.e. it solves the N+1 query problem.

Reply:
```
app_1_3c8c6cb7e593 | (0.001) SELECT "core_organisation"."id", "core_organisation"."created", "core_organisation"."updated", "core_organisation"."name", "core_organisation"."short_name", "core_organisation"."slug", "core_organisation"."summary", "core_organisation"."kind_id", "core_organisation"."started", "core_organisation"."ended", "core_organisation"."seats", "core_organisation"."show_attendance", "core_organisation"."_summary_rendered" FROM "core_organisation" WHERE "core_organisation"."slug" = 'national-assembly'; args=(u'national-assembly',)
app_1_3c8c6cb7e593 | (0.021) SELECT DISTINCT "core_person"."id", "core_person"."created", "core_person"."updated", "core_person"."title", "core_person"."legal_name", "core_person"."slug", "core_person"."gender", "core_person"."date_of_birth", "core_person"."date_of_death", "core_person"."summary", "core_person"."email", "core_person"."hidden", "core_person"."can_be_featured", "core_person"."biography", "core_person"."national_identity", "core_person"."family_name", "core_person"."given_name", "core_person"."additional_name", "core_person"."honorific_prefix", "core_person"."honorific_suffix", "core_person"."sort_name", "core_person"."_biography_rendered", "core_person"."_summary_rendered" FROM "core_person" INNER JOIN "core_position" ON ( "core_person"."id" = "core_position"."person_id" ) WHERE ("core_person"."hidden" = false AND "core_position"."id" IN (SELECT U0."id" FROM "core_position" U0 WHERE (U0."sorting_start_date" <= '2020-05-26' AND (U0."sorting_end_date_high" >= '2020-05-26' OR U0."end_date" = '') AND U0."organisation_id" = 70))) ORDER BY "core_person"."sort_name" ASC; args=(False, '2020-05-26', '2020-05-26', '', 70)
app_1_3c8c6cb7e593 | (0.003) SELECT "core_alternativepersonname"."id", "core_alternativepersonname"."created", "core_alternativepersonname"."updated", "core_alternativepersonname"."person_id", "core_alternativepersonname"."alternative_name", "core_alternativepersonname"."name_to_use", "core_alternativepersonname"."start_date", "core_alternativepersonname"."end_date", "core_alternativepersonname"."note", "core_alternativepersonname"."family_name", "core_alternativepersonname"."given_name", "core_alternativepersonname"."additional_name", "core_alternativepersonname"."honorific_prefix", "core_alternativepersonname"."honorific_suffix" FROM "core_alternativepersonname" WHERE "core_alternativepersonname"."person_id" IN (1542, 14995, ...)
app_1_3c8c6cb7e593 | (0.051) SELECT "core_position"."id", "core_position"."created", "core_position"."updated", "core_position"."person_id", "core_position"."organisation_id", "core_position"."place_id", "core_position"."title_id", "core_position"."subtitle", "core_position"."category", "core_position"."note", "core_position"."start_date", "core_position"."end_date", "core_position"."sorting_start_date", "core_position"."sorting_end_date", "core_position"."sorting_start_date_high", "core_position"."sorting_end_date_high", "core_organisation"."id", "core_organisation"."created", "core_organisation"."updated", "core_organisation"."name", "core_organisation"."short_name", "core_organisation"."slug", "core_organisation"."summary", "core_organisation"."kind_id", "core_organisation"."started", "core_organisation"."ended", "core_organisation"."seats", "core_organisation"."show_attendance", "core_organisation"."_summary_rendered" FROM "core_position" INNER JOIN "core_positiontitle" ON ( "core_position"."title_id" = "core_positiontitle"."id" ) INNER JOIN "core_organisation" ON ( "core_position"."organisation_id" = "core_organisation"."id" ) INNER JOIN "core_organisationkind" ON ( "core_organisation"."kind_id" = "core_organisationkind"."id" ) WHERE ("core_position"."sorting_start_date" <= '2020-05-26' AND ("core_position"."sorting_end_date_high" >= '2020-05-26' OR "core_position"."end_date" = '') AND "core_positiontitle"."slug" = 'member' AND "core_organisationkind"."slug" = 'party' AND "core_position"."person_id" IN (1542, 14995, ...)
app_1_3c8c6cb7e593 | (0.006) SELECT "core_contact"."id", "core_contact"."created", "core_contact"."updated", "core_contact"."kind_id", "core_contact"."value", "core_contact"."note", "core_contact"."source", "core_contact"."preferred", "core_contact"."content_type_id", "core_contact"."object_id" FROM "core_contact" INNER JOIN "core_contactkind" ON ( "core_contact"."kind_id" = "core_contactkind"."id" ) WHERE ("core_contactkind"."slug" IN ('cell', 'phone') AND "core_contact"."object_id" IN (10259, 10260, ...)
app_1_3c8c6cb7e593 | (0.010) SELECT "core_contact"."id", "core_contact"."created", "core_contact"."updated", "core_contact"."kind_id", "core_contact"."value", "core_contact"."note", "core_contact"."source", "core_contact"."preferred", "core_contact"."content_type_id", "core_contact"."object_id" FROM "core_contact" INNER JOIN "core_contactkind" ON ( "core_contact"."kind_id" = "core_contactkind"."id" ) WHERE ("core_contactkind"."slug" IN ('email') AND "core_contact"."object_id" IN (10259, 10260,...)
```

Reply:
ah cool - something to watch for is if this is used for an organisation like the ANC that has more members than our national assembly does at any one time - then that `IN` can get out of hand.